### PR TITLE
Document Template Admin Page | Features/psp 3945

### PR DESCRIPTION
**Pull request by FuriousLlama**
image

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -19,16 +19,16 @@
...
       "args": [],
-      "cwd": "${workspaceFolder}/backend/api",
+      "cwd": "${workspaceRoot}/source/backend/api",
       "stopAtEntry": false,
       "env": {
         "ASPNETCORE_ENVIRONMENT": "Local",
         "ASPNETCORE_URLS": "http://*:5000;https://*:433"
       },
       "sourceFileMap": {
-        "/Views": "${workspaceFolder}/backend/Views"
```
Comment (devinleighsmith) [inline .vscode/launch.json:31]: thanks for this.

File: source/backend/api/Areas/Documents/DocumentRelationshipController.cs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+        #region Constructors
+
+        /// <summary>
+        /// Creates a new instance of a DocumentRelationshipController class.
+        /// </summary>
+        /// <param name="documentService"></param>
+        /// <param name="documentActivityService"></param>
+        /// <param name="mapper"></param>
+        public DocumentRelationshipController(IDocumentService documentService, IDocumentActivityService documentActivityService, IMapper mapper)
```
Comment (devinleighsmith) [inline source/backend/api/Areas/Documents/DocumentRelationshipController.cs:41]: I like this change, it makes sense.

File: source/backend/api/Services/AcquisitionFileService.cs [hunk 3]
```diff
@@ -154,6 +154,11 @@ public PimsAcquisitionFile UpdateProperties(PimsAcquisitionFile acquisitionFile)
             return _acqFileRepository.GetById(acquisitionFile.Id);
         }
 
+        private static void UpdateFileNumber(PimsAcquisitionFile acquisitionFile)
```
Comment (devinleighsmith) [inline source/backend/api/Services/AcquisitionFileService.cs:157]: This doesn't seem related to these changes, should this be here?

Comment (devinleighsmith) [reply source/backend/api/Services/AcquisitionFileService.cs:157]: just a warning about alpha-order maybe?

Comment (devinleighsmith) [reply source/backend/api/Services/AcquisitionFileService.cs:157]: or private static vs private perhaps.

File: source/backend/api/Services/DocumentActivityService.cs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+using Pims.Api.Models.Concepts;
+using Pims.Dal.Entities;
+using Pims.Dal.Helpers.Extensions;
+using Pims.Dal.Repositories;
+using Pims.Dal.Security;
+
+namespace Pims.Api.Services
+{
+    /// <summary>
+    /// DocumentService implementation provides document managing capabilities.
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentActivityService.cs:16]: nit: copy pasta

File: source/backend/api/Services/IDocumentActivityService.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Pims.Api.Models;
+using Pims.Api.Models.Concepts;
+using Pims.Dal.Entities;
+
+namespace Pims.Api.Services
+{
+    /// <summary>
+    /// IDocumentSIDocumentActivityServicervice interface, defines the functionality for document services.
```
Comment (devinleighsmith) [inline source/backend/api/Services/IDocumentActivityService.cs:10]: nit: typo

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        public DocumentActivityTemplateRepository(PimsContext dbContext, ClaimsPrincipal user, ILogger<DocumentActivityRepository> logger)
+            : base(dbContext, user, logger)
+        {
+        }
+        #endregion
+
+        #region Methods
+
+        /// <summary>
+        /// Get a list of all the document relationships for a given document.
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:34]: nit: comment may be copy/pasta

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                    .ThenInclude(d => d.DocumentStatusTypeCodeNavigation)
+                .Include(ad => ad.Document)
+                    .ThenInclude(d => d.DocumentType)
+                .Where(ad => ad.DocumentId == documentId)
+                .AsNoTracking()
+                .ToList();
+        }
+
+        /// <summary>
+        /// Get a list of all the document relationships for a given activity template.
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:50]: nit: might be all the comments in this repo still refer to document relationships when this repo is specific to templates.

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        /// <param name="activityTemplateDocument"></param>
+        /// <returns></returns>
+        public PimsActivityTemplateDocument Add(PimsActivityTemplateDocument activityTemplateDocument)
+        {
+            activityTemplateDocument.ThrowIfNull(nameof(activityTemplateDocument));
+
+            var newEntry = this.Context.PimsActivityTemplateDocuments.Add(activityTemplateDocument);
+            if (newEntry.State == EntityState.Added)
+            {
+                return newEntry.Entity;
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:77]: This is interesting, seems more efficient then doing another database roundtrip by calling the fetch method.

File: source/backend/api/Services/AcquisitionFileService.cs (hunk 3 above)
Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:157]: yep, just a warning fix on the static

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs (hunk 8 above)
Comment (FuriousLlama) [reply source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:77]: Yes it does if the commit is executed, otherwise no.

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs (hunk 7 above)
Comment (FuriousLlama) [reply source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:50]: this one is pretty close, at the end of the doc there is the specification that this is for an activity template

File: source/frontend/package-lock.json [hunk 9]
```diff
@@ -156,6 +156,19 @@
         "multicast-dns": "cli.js"
       }
     },
+    "node_modules/pkg-up/node_modules/locate-path": {
```
[the remark this answers is not in the document]
Comment (FuriousLlama) [reply source/frontend/package-lock.json:159]: not sure, really. I ran the install and this one popped up

File: source/frontend/src/features/admin/document-template/DocumentTemplateManagementView.tsx [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    <ListPage>
+      <Scrollable>
+        <LoadingBackdrop show={props.isLoading} />
+
+        <StyledPageHeader>PIMS Document Template Management</StyledPageHeader>
+        <Section>
+          <Row>
+            <Col xs="auto">Activity Type:</Col>
+            <Col xs="auto">
+              <Form.Group aria-label="Default select example">
```
Comment (devinleighsmith) [inline source/frontend/src/features/admin/document-template/DocumentTemplateManagementView.tsx:39]: default select example?

File: source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx [hunk 11]
```diff
@@ -62,30 +62,45 @@ export const DocumentUploadContainer: React.FunctionComponent<IDocumentUploadCon
...
+    [retrieveDocumentTypeMetadata],
+  );
+
   useEffect(() => {
     const fetch = async () => {
       const axiosResponse = await retrieveDocumentTypes();
       if (axiosResponse && isMounted()) {
-        setDocumentTypes(axiosResponse);
+        if (props.relationshipType === DocumentRelationshipType.TEMPLATES) {
+          setDocumentTypes(axiosResponse.filter(x => x.documentType === 'CDOGS Template'));
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:94]: nit: would be nice to use a constant for this.

Comment (FuriousLlama) [reply source/frontend/src/features/documents/documentUpload/DocumentUploadContainer.tsx:94]: sure thing.

File: source/frontend/src/features/documents/documentUpload/DocumentUploadForm.tsx [hunk 12]
```diff
@@ -100,7 +101,7 @@ const DocumentUploadForm: React.FunctionComponent<IDocumentUploadFormProps> = pr
             <SectionField label="Document type" labelWidth="3" className="pb-2">
               <Select
                 data-testid="document-type"
-                placeholder="Select Document type"
+                placeholder={documentTypes.length > 1 ? 'Select Document type' : undefined}
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/documentUpload/DocumentUploadForm.tsx:104]: this doesn't end up printing 'undefined' does it?

Comment (FuriousLlama) [reply source/frontend/src/features/documents/documentUpload/DocumentUploadForm.tsx:104]: It does not. If placeholder is set to undefined, then it selects the first option

File: source/frontend/src/features/documents/list/DocumentListView.tsx [hunk 13]
```diff
@@ -34,12 +35,33 @@ export const DocumentListView: React.FunctionComponent<IDocumentListViewProps> =
...
   const [filters, setFilters] = React.useState<IDocumentFilter>(
     defaultFilters ?? defaultDocumentFilter,
   );
 
+  const { retrieveDocumentTypes } = useDocumentProvider();
+
+  useEffect(() => {
+    const fetch = async () => {
+      const axiosResponse = await retrieveDocumentTypes();
+      console.log(axiosResponse, props.relationshipType);
```
Comment (devinleighsmith) [inline source/frontend/src/features/documents/list/DocumentListView.tsx:51]: console.log

**Review by devinleighsmith — COMMENTED** [review]
I'm just going to check this out really quick and run some tests and then will finish up this review.

File: source/backend/api/Services/AcquisitionFileService.cs (hunk 3 above)
Comment (asanchezr) [reply source/backend/api/Services/AcquisitionFileService.cs:155]: This change is no longer relevant so I would revert. Not sure how you have that since my PR (merged a few days ago) removed that method (because number generation was moved to repository). Maybe you need to rebase?

Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:155]: good to know. I will check double check that is no longer necessary

File: source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs [hunk 14]
```diff
@@ -1,16 +1,19 @@
...
+using Pims.Dal.Security;
 using Swashbuckle.AspNetCore.Annotations;
 
 namespace Pims.Api.Controllers
 {
     /// <summary>
     /// DocumentGenerationController class, provides endpoints to handle document generation requests.
     /// </summary>
-    //[Authorize]
+    [Authorize]
```
Comment (asanchezr) [inline source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: I see you enabled authorization here but still relevant claims on specific methods below are commented out. Shouldn't they be enabled as well?

Comment (FuriousLlama) [reply source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: I think Devin has a PR for those claims types. Until that is merged the methods below cannot be commented out.

File: source/backend/api/Services/DocumentActivityService.cs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+            this.User.ThrowIfNotAuthorized(Permissions.DocumentAdmin);
+
+            DocumentUploadResponse uploadResult = await documentService.UploadDocumentAsync(uploadRequest);
+
+            DocumentUploadRelationshipResponse relationshipResponse = new DocumentUploadRelationshipResponse()
+            {
+                UploadResponse = uploadResult,
+            };
+
+            if (uploadResult.Document.Id != 0)
```
Comment (devinleighsmith) [inline source/backend/api/Services/DocumentActivityService.cs:97]: I consistently get an NPE here when uploading a CDOGS template document. The uploadResult.Document is null, only the uploadResult.DocumentExternalResult is populated.

**Review by devinleighsmith — APPROVED** [review]

File: source/backend/api/Services/DocumentActivityService.cs (hunk 15 above)
Comment (FuriousLlama) [reply source/backend/api/Services/DocumentActivityService.cs:97]: uhm that didn't show up on my side. I'll double check

Comment (devinleighsmith) [reply source/backend/api/Services/DocumentActivityService.cs:97]: This appears to be because the repo is returning an error and this service is not handling the error appropriately.

Comment (FuriousLlama) [reply source/backend/api/Services/DocumentActivityService.cs:97]: It worked on my side. Looks like your mayan api is not returning a good result. I will make this check more robust

File: source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs (hunk 14 above)
Comment (asanchezr) [reply source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: Fair enough, but still confused as to why your PR cannot uncomment the methods below but your new code can do:


```csharp
 this.User.ThrowIfNotAuthorized(Permissions.DocumentView);
```

Aren't they the same document permissions?

Comment (asanchezr) [reply source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: Like here in Line 44 `//[HasPermission(Permissions.DocumentView)]`

File: source/frontend/src/constants/documentType.ts [hunk 16]
```diff
@@ -0,0 +1,3 @@
+export enum DocumentTypeName {
+  CDOGS = 'CDOGS Template',
```
Comment (devinleighsmith) [inline source/frontend/src/constants/documentType.ts:2]: My fault on this one, one of my stories changed CDOGS Template to CDOGS template.

File: source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs (hunk 14 above)
Comment (devinleighsmith) [reply source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: My PR shouldn't affect any of the claims in this sprint. It should allow a developer in the next sprint to allow document generation to normal (non-admin) users.

Comment (FuriousLlama) [reply source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs:16]: Probably both need to be changed. The new permission is DocumentGenerate.

File: source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        /// Deletes the passed document activity template to the database.
+        /// </summary>
+        /// <param name="activityTemplateDocument"></param>
+        /// <returns></returns>
+        public bool Delete(PimsActivityTemplateDocument activityTemplateDocument)
+        {
+            if (activityTemplateDocument == null)
+            {
+                throw new ArgumentNullException(nameof(activityTemplateDocument), "activityTemplateDocument cannot be null.");
+            }
```
Comment (asanchezr) [inline source/backend/dal/Repositories/DocumentActivityTemplateRepository.cs:95]: nit: we already have an extension method to reduce boilerplate:

```csharp
activityTemplateDocument.ThrowIfNull(nameof(activityTemplateDocument));
```

File: source/backend/tests/unit/api/Controllers/Activities/AcquisitionControllerTest.cs [hunk 18]
```diff
@@ -334,7 +334,7 @@ public void UpdateActivityProperty_BadRequest()
             service.Setup(m => m.UpdateActivityAcquisitionProperties(It.IsAny<PimsActivityInstance>())).Returns(activity);
 
             // Act
-            Action act =  () => controller.UpdateActivityProperties(FileType.Unknown, mapper.Map<ActivityInstanceModel>(activity));
+            Action act = () => controller.UpdateActivityProperties(FileType.Unknown, mapper.Map<ActivityInstanceModel>(activity));
```
Comment (asanchezr) [inline source/backend/tests/unit/api/Controllers/Activities/AcquisitionControllerTest.cs:337]: Since you are modifying this file - could you please rename it to match the class name?
image
